### ISSUE TITLE
[Omniscia audit] Unwarranted Halt of Execution For Self-Orders

### ISSUE THREAD

**Issue by hiroshi-yamamoto-dublr** [dublr/dublr#4]
## DUB-03M: Unwarranted Halt of Execution For Self-Orders

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) | <img class="o-severity o-medium" src="https://omniscia.io/report-assets/medium.png" /> | [Dublr.sol:L420](https://github.com/dublr/dublr/blob/b84ea8edd8c1ab6a3f1abbc34467fa2ce4b21526/contracts/main/Dublr/Dublr.sol#L420) |

### Description:

The `buy` function will cease operation if the sell orders have been consumed up to the point that one's own sell order has been reached. This is a significant issue as it will be fairly common for sellers to buy lower-price orders as soon as they appear to artificially pump the price of their tokens.

### Impact:

The current implementation does not take transaction ordering into account and is prone to fail transactions that should otherwise be properly executed.

### Example:

```solidity
for (uint256 numSellOrdersBought = 0;
        // If buyingEnabled is false, skip over the buying stage (minting is still enabled unless
        // mintingEnabled is also false). This allows exchange function to be shut down if necessary
        // without affecting minting.
        buyingEnabled
        // Iterate through orders in increasing order of priceETHPerDUBLR_x1e9, until we run out of ETH,
        // or until we run out of orders.
        && buyOrderRemainingETHWEI > 0 && orderBook.length > 0; ) {
        
    // Find the lowest-priced order (this is a memory copy, because heapRemove(0) may be called below)
    Order memory sellOrder = orderBook[0];
    // If you're trying to buy your own sell order, you need to cancel your sell order first
    require (sellOrder.seller != buyer, "Can't buy own sell order");
```

### Recommendation:

We advise the `buy` order to instead remove the caller's order from the order book if it exists and re-introduce it after the buy loop has concluded, ensuring that sellers can properly purchase orders below them without fear of their transaction failing.

**Comment by hiroshi-yamamoto-dublr** [comment]
I misunderstood the recommendation. Fixed in [e9a1a92](https://github.com/dublr/dublr/commit/e9a1a92f1845d812511e37ee12219e4befdecf0c).

**Comment by hiroshi-yamamoto-dublr** [comment]
Further simplified in 66666e6.
